### PR TITLE
Fixes #271: Use parametrize markers params for getting example_kwargs

### LINKED ISSUE #271: Running gherkin reporter on pytest-bdd tests raises KeyError
Running below test:
```
pytest -k 'test_parametrized' --gherkin-terminal-reporter --gherkin-terminal-reporter-expanded
```

Raises following error:
```
Feature:
    Scenario: Parametrized given, when, thens
INTERNALERROR> Traceback (most recent call last):
INTERNALERROR>   File "/Users/milosz.sliwinski/.virtualenvs/pytest-bdd/lib/python2.7/site-packages/_pytest/main.py", line 96, in wrap_session
INTERNALERROR>     session.exitstatus = doit(config, session) or 0
INTERNALERROR>   File "/Users/milosz.sliwinski/.virtualenvs/pytest-bdd/lib/python2.7/site-packages/_pytest/main.py", line 131, in _main
INTERNALERROR>     config.hook.pytest_runtestloop(session=session)
INTERNALERROR>   File "/Users/milosz.sliwinski/.virtualenvs/pytest-bdd/lib/python2.7/site-packages/_pytest/vendored_packages/pluggy.py", line 724, in __call__
INTERNALERROR>     return self._hookexec(self, self._nonwrappers + self._wrappers, kwargs)
INTERNALERROR>   File "/Users/milosz.sliwinski/.virtualenvs/pytest-bdd/lib/python2.7/site-packages/_pytest/vendored_packages/pluggy.py", line 338, in _hookexec
INTERNALERROR>     return self._inner_hookexec(hook, methods, kwargs)
INTERNALERROR>   File "/Users/milosz.sliwinski/.virtualenvs/pytest-bdd/lib/python2.7/site-packages/_pytest/vendored_packages/pluggy.py", line 333, in <lambda>
INTERNALERROR>     _MultiCall(methods, kwargs, hook.spec_opts).execute()
INTERNALERROR>   File "/Users/milosz.sliwinski/.virtualenvs/pytest-bdd/lib/python2.7/site-packages/_pytest/vendored_packages/pluggy.py", line 596, in execute
INTERNALERROR>     res = hook_impl.function(*args)
INTERNALERROR>   File "/Users/milosz.sliwinski/.virtualenvs/pytest-bdd/lib/python2.7/site-packages/_pytest/main.py", line 152, in pytest_runtestloop
INTERNALERROR>     item.config.hook.pytest_runtest_protocol(item=item, nextitem=nextitem)
INTERNALERROR>   File "/Users/milosz.sliwinski/.virtualenvs/pytest-bdd/lib/python2.7/site-packages/_pytest/vendored_packages/pluggy.py", line 724, in __call__
INTERNALERROR>     return self._hookexec(self, self._nonwrappers + self._wrappers, kwargs)
INTERNALERROR>   File "/Users/milosz.sliwinski/.virtualenvs/pytest-bdd/lib/python2.7/site-packages/_pytest/vendored_packages/pluggy.py", line 338, in _hookexec
INTERNALERROR>     return self._inner_hookexec(hook, methods, kwargs)
INTERNALERROR>   File "/Users/milosz.sliwinski/.virtualenvs/pytest-bdd/lib/python2.7/site-packages/_pytest/vendored_packages/pluggy.py", line 333, in <lambda>
INTERNALERROR>     _MultiCall(methods, kwargs, hook.spec_opts).execute()
INTERNALERROR>   File "/Users/milosz.sliwinski/.virtualenvs/pytest-bdd/lib/python2.7/site-packages/_pytest/vendored_packages/pluggy.py", line 595, in execute
INTERNALERROR>     return _wrapped_call(hook_impl.function(*args), self.execute)
INTERNALERROR>   File "/Users/milosz.sliwinski/.virtualenvs/pytest-bdd/lib/python2.7/site-packages/_pytest/vendored_packages/pluggy.py", line 253, in _wrapped_call
INTERNALERROR>     return call_outcome.get_result()
INTERNALERROR>   File "/Users/milosz.sliwinski/.virtualenvs/pytest-bdd/lib/python2.7/site-packages/_pytest/vendored_packages/pluggy.py", line 279, in get_result
INTERNALERROR>     _reraise(*ex)  # noqa
INTERNALERROR>   File "/Users/milosz.sliwinski/.virtualenvs/pytest-bdd/lib/python2.7/site-packages/_pytest/vendored_packages/pluggy.py", line 264, in __init__
INTERNALERROR>     self.result = func()
INTERNALERROR>   File "/Users/milosz.sliwinski/.virtualenvs/pytest-bdd/lib/python2.7/site-packages/_pytest/vendored_packages/pluggy.py", line 596, in execute
INTERNALERROR>     res = hook_impl.function(*args)
INTERNALERROR>   File "/Users/milosz.sliwinski/.virtualenvs/pytest-bdd/lib/python2.7/site-packages/_pytest/runner.py", line 66, in pytest_runtest_protocol
INTERNALERROR>     runtestprotocol(item, nextitem=nextitem)
INTERNALERROR>   File "/Users/milosz.sliwinski/.virtualenvs/pytest-bdd/lib/python2.7/site-packages/_pytest/runner.py", line 79, in runtestprotocol
INTERNALERROR>     reports.append(call_and_report(item, "call", log))
INTERNALERROR>   File "/Users/milosz.sliwinski/.virtualenvs/pytest-bdd/lib/python2.7/site-packages/_pytest/runner.py", line 137, in call_and_report
INTERNALERROR>     hook.pytest_runtest_logreport(report=report)
INTERNALERROR>   File "/Users/milosz.sliwinski/.virtualenvs/pytest-bdd/lib/python2.7/site-packages/_pytest/vendored_packages/pluggy.py", line 724, in __call__
INTERNALERROR>     return self._hookexec(self, self._nonwrappers + self._wrappers, kwargs)
INTERNALERROR>   File "/Users/milosz.sliwinski/.virtualenvs/pytest-bdd/lib/python2.7/site-packages/_pytest/vendored_packages/pluggy.py", line 338, in _hookexec
INTERNALERROR>     return self._inner_hookexec(hook, methods, kwargs)
INTERNALERROR>   File "/Users/milosz.sliwinski/.virtualenvs/pytest-bdd/lib/python2.7/site-packages/_pytest/vendored_packages/pluggy.py", line 333, in <lambda>
INTERNALERROR>     _MultiCall(methods, kwargs, hook.spec_opts).execute()
INTERNALERROR>   File "/Users/milosz.sliwinski/.virtualenvs/pytest-bdd/lib/python2.7/site-packages/_pytest/vendored_packages/pluggy.py", line 596, in execute
INTERNALERROR>     res = hook_impl.function(*args)
INTERNALERROR>   File "/Users/milosz.sliwinski/workspace/pytest-bdd/pytest_bdd/gherkin_terminal_reporter.py", line 104, in pytest_runtest_logreport
INTERNALERROR>     step_name = self._format_step_name(step['name'], **report.scenario['example_kwargs'])
INTERNALERROR>   File "/Users/milosz.sliwinski/workspace/pytest-bdd/pytest_bdd/gherkin_terminal_reporter.py", line 122, in _format_step_name
INTERNALERROR>     param_value = example_kwargs[param_name]
INTERNALERROR> KeyError: u'start'
```

### PR DESCRIPTION
**Pull request by sliwinski-milosz**
Fixes https://github.com/pytest-dev/pytest-bdd/issues/271

User can parametrize scenario by using examples specified in feature files. In that case pytest `parametrize` marker is updated with examples parameters.

We also provide a way to parametrize scenario on the python side by using `pytest.mark.parametrize`. https://pytest-bdd.readthedocs.io/en/latest/#combine-scenario-outline-and-pytest-parametrization 

The problem with gherkin reporting is that when we are getting `self.example_kwargs`, we are using `scenario.get_example_params()` method that takes into account only args that are defined inside feature file (examples), and skips `pytest.mark.parametrize` parameters.

For gherkin reporting we need all parameters: those specified in feature files and those specified by using `pytest.mark.parametrize` decorator.

This pull requests uses `parametrize` marker to get all args.

### CODE REVIEW

File: pytest_bdd/utils.py [hunk 1]
```diff
@@ -110,3 +110,33 @@ def get_markers_args_using_iter_markers(node, mark_name):
...
+    params = []
+    for i in range(0, len(parametrize_args), 2):
+        params.append({
+            'names': _get_param_names(parametrize_args[i]),
+            'values': parametrize_args[i+1]
+        })
+    return params
+
+
+def _get_param_names(names):
```
Comment (youtux) [inline pytest_bdd/utils.py]: maybe this function can be called `coerce_list`

File: tests/feature/test_outline.py [hunk 2]
```diff
@@ -168,7 +168,7 @@ def should_have_left_fruits(start_fruits, start, eat, left, fruits):
 def test_outlined_feature(request):
     assert get_parametrize_markers_args(request.node) == (
         ['start', 'eat', 'left'],
-        [[12, 5.0, '7'], [5, 4.0, '1']],
+        [[12, 5.0, '7'], [5, 4.0, '1'], [4, 2.0, '2']],
```
Comment (youtux) [inline tests/feature/test_outline.py]: what is this change for?

File: tests/feature/test_parametrized.py [hunk 3]
```diff
@@ -14,6 +14,17 @@ def test_parametrized(request, start, eat, left):
...
 
 
+@pytest.mark.parametrize(
+    'start', [12]
+)
+@scenario(
+    'parametrized.feature',
+    'Parametrized given - single param',
+)
+def test_parametrized_single_param(request, start):
```
Comment (youtux) [inline tests/feature/test_parametrized.py]: what is this test testing?

File: pytest_bdd/utils.py (hunk 1 above)
Comment (sliwinski-milosz) [reply pytest_bdd/utils.py]: It makes sense, renamed

Comment (sliwinski-milosz) [reply pytest_bdd/utils.py]: On the other hand `_coerce_list` may return `tuple` -> maybe we can find even better name like `_coerce_list_if_single`

File: pytest_bdd/utils.py [hunk 4]
```diff
@@ -110,3 +110,33 @@ def get_markers_args_using_iter_markers(node, mark_name):
...
+        })
+    return params
+
+
+def _coerce_list(names):
+    if not isinstance(names, (tuple, list)):
+        # As pytest.mark.parametrize has only one param name,
+        # it is not returned as a list. Convert it to list:
+        names = [names]
+    return names
```
Comment (youtux) [inline pytest_bdd/utils.py]: you can then `return list(names)` here, this will make it a list in either case.

**Comment by sliwinski-milosz** [comment]
Please before merging, let me add similar test as in fix of unicode steps (but to do that easily unicode steps fix needs to be merged)

**Comment by sliwinski-milosz** [comment]
I added test and I made test function bit more generic.

However I still think that it would be good to firstly merge the fix for CI tests :)

**Comment by sliwinski-milosz** [comment]
Please note that "work in progress" label can be removed. This pull request is ready for review.

File: tests/utils.py [hunk 5]
```diff
@@ -6,13 +6,22 @@ def get_test_filepath(filepath):
     return os.path.join(curr_file_dirpath, filepath)
 
 
+def get_filename_without_ext(path):
+    filename = os.path.basename(path)
+    return os.path.splitext(filename)[0]
+
+
 def prepare_feature_and_py_files(testdir, feature_file, py_file):
```
Comment (youtux) [inline tests/utils.py]: This function was used only once, and it's implementation was tied to the caller. I would say to remove it and let the caller `test_scenario_in_expanded_mode` test manage the resources (and let the test be in the pytest-bdd style, not a normal test)

File: tests/feature/parametrized.feature [hunk 6]
```diff
@@ -2,3 +2,7 @@ Scenario: Parametrized given, when, thens
     Given there are <start> cucumbers
     When I eat <eat> cucumbers
     Then I should have <left> cucumbers
+
+
+Scenario: Parametrized given - single param
```
Comment (youtux) [inline tests/feature/parametrized.feature]: what is this test asserting? There is no "then" step

File: tests/feature/test_gherkin_terminal_reporter.py [hunk 7]
```diff
@@ -349,18 +349,18 @@ def output_output_must_contain_parameters_values(test_execution, gherkin_scenari
...
+    result = testdir.runpytest(
+        '%s::%s' % (py_filename, name),
         '--gherkin-terminal-reporter',
         '--gherkin-terminal-reporter-expanded',
         '-vv',
     )
-
-    ghe = test_execution['gherkin']
-    ghe.assert_outcomes(passed=1)
+    result.assert_outcomes(passed=1)
```
Comment (youtux) [inline tests/feature/test_gherkin_terminal_reporter.py]: we are asserting that the test pass, but not the behaviour of the terminal report, which is the whole point of this test.

Comment (sliwinski-milosz) [reply tests/feature/test_gherkin_terminal_reporter.py]: The problem that this pull request fixes is that expanded gherkin report was crashing while reporting `test_parametrized` test.
So this assertion just checks if test was able to run or if there was a crash.

File: tests/feature/parametrized.feature (hunk 6 above)
Comment (sliwinski-milosz) [reply tests/feature/parametrized.feature]: You already have asked that question :). Shall we add comment somewhere in the code explaining what this test is testing?

https://github.com/pytest-dev/pytest-bdd/pull/272#discussion_r236071403

File: tests/utils.py (hunk 5 above)
Comment (sliwinski-milosz) [reply tests/utils.py]: I will rewrite it to pytest-bdd style.

I added `prepare_feature_and_py_files` to be able to reuse feature files that are already in other tests, to test them in expanded mode.

I will remove it, and I'll create separate tests for gherkin reporter. That means that we may have some duplication but it is safer approach.

Comment (sliwinski-milosz) [reply tests/utils.py]: Tests has been rewritten to pytest-bdd style.

**Comment by sliwinski-milosz** [comment]
I think this PR is also ready for review :)
